fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1417,7 +1421,11 @@ def _remove_shard_bwd_hook(fsdp_module: FullyShardedDataParallel) -> None:
                 _remove_shard_bwd_hook(m)
                 m._pre_backward_hook_has_run = False
                 if any(p.requires_grad for p in m.parameters()):
-                    if m._has_params:
+                    # Check if the module has params and if any of them has
+                    # the `requires_grad` field set. If `requires_grad=False` for
+                    # all the params, the post_backward hook will not fire and the
+                    # state will remain in `TrainingState.BACKWARD_PRE`.
+                    if m._has_params and any([p.requires_grad for p in m.params]):
```

Comment:
cc @zhaojuanmao I had to add this condition along with the one above for ensuring that we check the BACKWARD_POST state if the root instance has param and they require grads. There is a use case where they may not. Let me know if this makes sense.

Reply:
same as above, perhaps checking _has_params is not needed?